packages/frontend/src/redux/slices/sign/index.js

```diff
@@ -1,4 +1,138 @@
+
 const SLICE_NAME = 'sign';
 
+export const SIGN_STATUS = {
+    IN_PROGRESS: 'in-progress',
+    NEEDS_CONFIRMATION: 'needs-confirmation',
+    RETRY_TRANSACTION: 'retry-tx',
+    SUCCESS: 'success',
+    ERROR: 'error'
+};
+
+export const RETRY_TX = {
+    INCREASE: 'increase',
+    GAS: {
+        DIFF: '25000000000000',
+        MAX: '300000000000000',
+    }
+};
+
+export const handleSignTransactions = createAsyncThunk(
+    `${SLICE_NAME}/handleSignTransactions`,
+    async (_, thunkAPI) => {
+        const { dispatch, getState } = thunkAPI;
+        let transactionsHashes;
+        const retryTxDirection = selectSignRetryTxDirection(getState());
+
+        const mixpanelName = `SIGN${retryTxDirection ? ` - RETRY - ${retryTxDirection}` : ''}`;
+        await Mixpanel.withTracking(mixpanelName,
+            async () => {
+                const transactions = selectSignTransactions(getState());
+                const accountId = selectAccountId(getState());
+
+                try {
+                    transactionsHashes = await wallet.signAndSendTransactions(transactions, accountId);
+                } catch (error) {
+                    if (error.message.includes('TotalPrepaidGasExceeded')) {
+                        Mixpanel.track('SIGN - too much gas detected');
+                    }
+
+                    dispatch(showCustomAlert({
+                        success: false,
+                        messageCodeHeader: 'error',
+                        messageCode: `reduxActions.${error.code}`,
+                        errorMessage: error.message
+                    }));
+                    throw error;
+                }
+            }
+        );
+
+        return transactionsHashes;
+    },
+    {
+        condition: (_, thunkAPI) => {
+            const { getState } = thunkAPI;
+            if (selectSignStatus(getState()) === SIGN_STATUS.IN_PROGRESS) {
+                return false;
+            }
+        }
+    }
+);
+
+export function addQueryParams(baseUrl, queryParams) {
+    const url = new URL(baseUrl);
+    for (let key in queryParams) {
+        const param = queryParams[key];
+        if(param) url.searchParams.set(key, param);
+    }
+    return url.toString();
+}
+
+export const increaseGasForTransactions = ({ transactions, retryTxDirection }) => {
+    transactions.forEach((t) => {
+        const oneFunctionCallAction = t.actions && t.actions.filter((a) => !!a.functionCall).length === 1;
+
+        t.actions && t.actions.forEach((a) => {
+            if(a.functionCall && a.functionCall.gas && retryTxDirection === RETRY_TX.INCREASE) {
+                oneFunctionCallAction
+                    ? a.functionCall.gas = new BN(RETRY_TX.GAS.MAX)
+                    : a.functionCall.gas = BN.min(
+                        new BN(RETRY_TX.GAS.MAX),
+                        a.functionCall.gas.add(new BN(RETRY_TX.GAS.DIFF))
+                    );
+            }
+        });
+    });
+    return transactions;
+};
```

Comment:
1. It doesn't make sense for `increaseGasForTransactions` to take a `retryTxDirection`.  If we ever call this method, we expect gas to be increased; :hocho: `retryTxDirection`

2. Please unravel these expressions and use early exits to remove indentation levels.  By doing so, you will also add nice convenient spaces to add comments that explain _why_ we are doing the things we are doing here for later.  For example, the reason why we don't always increase every action's gas to GAS.MAX is an important one, but someone would have to guess why based on the current code.  Let's be nice to our future selves and future new team members who need to read this later.

Something like e.g.:
```javascript
export const increaseGasForTransactions = ({ transactions, retryTxDirection }) => {
    transactions.forEach((t) => {
        const oneFunctionCallAction = t.actions && t.actions.filter((a) => !!a.functionCall).length === 1;

        t.actions && t.actions.forEach((a) => {
            if (!(a.functionCall && a.functionCall.gas)) { return false; }

            if (oneFunctionCallAction) {
                // If we only have a single functionCall type action, it is probably safe to immediately try GAS.MAX for that one action
                a.functionCall.gas = new BN(RETRY_TX.GAS.MAX);
                return;
            }

            // If there are more than one functionCall type actions, we will try to incrementally increase
            // gas allocated -- we don't increase straight to GAS.MAX because it is highly likely that we will
            // attach too much gas and encounter a `TotalPrepaidGasExceeded` type of error
            a.functionCall.gas = BN.min(
                new BN(RETRY_TX.GAS.MAX),
                a.functionCall.gas.add(new BN(RETRY_TX.GAS.DIFF))
            );
        });
    });
    return transactions;
};
```

Reply:
Done. Thanks for the example.